### PR TITLE
Introduce bitMatrix to replace mat.Dense

### LINKED ISSUE #22: Proposal: use light-weight bitmask instead of gonums's mat.Dense
This float64 slice
https://github.com/lazyledger/rsmt2d/blob/ef1d6c54461edaa428b5720835b6b522c18defff/extendeddatacrossword.go#L52-L63
is used to initialize a dense matrix: https://github.com/lazyledger/rsmt2d/blob/ef1d6c54461edaa428b5720835b6b522c18defff/extendeddatacrossword.go#L82

It looks like this could be vastly improved by using some form of 2d-bitmask to look up if a cell in a row or column is present or not, or if a row / column is incomplete. So instead of 64 bits per element, we could reduce that to 1 bit per element.

We can either look for a small library out there to use or we can write this our selves.

### PR DESCRIPTION
**Pull request by liamsi**
This PR introduces a structure that replaces mat.Dense and gets rid of the gonum dependency.
~~Hence, it is half the story of~~ closes #22.

It could use some light fuzzing but given that this is essentially a simple bit-mask I don't think that is urgent.

~~Can open a separate PR to actually use the structure, or, replace mat.Dense in here.~~

### CODE REVIEW

File: bit_matrix.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
+package rsmt2d
+
+import "fmt"
+
+type bitMatrix struct {
```
Comment (liamsi) [inline bit_matrix.go:5]: Maybe, I should note that the thing is not go-rountine-safe. That doesn't matter at all right now but I know that @evan-forbes experimented with paralleization in https://github.com/lazyledger/rsmt2d/issues/5 already.

**Comment by liamsi** [comment]
As requested, this PR also replaces gonum now. It would have been nice to measure the impact with benchmarks but we already know the ballpark of improvment (see #22).

File: bit_matrix.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	bits := squareSize * squareSize
+	return bitMatrix{mask: make([]uint64, (bits+63)/64), squareSize: squareSize}
+}
+
+// i = rowIndex*squareSize+colIdx
+func (bm bitMatrix) SetFlat(i int) {
+	bm.mask[i/64] |= uint64(1) << uint(i%64)
+}
+
+func (bm bitMatrix) Get(row, col int) bool {
```
Comment (evan-forbes) [inline bit_matrix.go:20]: Do these methods have to be exported?

Comment (liamsi) [reply bit_matrix.go:20]: Not really, no. Everything is in a single flat root package here anyways and even the type itself is not exported `bitMatrix`. Both the type and these methods would need to be exported if this was moved into a separaste go package. 
I only exported all these methods so it is clear which are meant to be used as an API, in other words, to communicate the surface of this if it was a library. The caller can currently bypass that anyways: they could directly mingle with the underlying `mask` field.

Comment (evan-forbes) [reply bit_matrix.go:20]: I can definitely dig that, nice

File: bit_matrix.go (hunk 1 above)
Comment (adlerjohn) [reply bit_matrix.go:5]: What's the alternative here? How much work would it be to make this thread-safe? Is there not a standard library for this that is thread-safe?

Comment (liamsi) [reply bit_matrix.go:5]: > How much work would it be to make this thread-safe?

Basically adding mutexes for accessing the underlying field. 

> Is there not a standard library for this that is thread-safe?

Not to my knowledge.

Also, we currently do not need this to be thread-safe as we aren't accessing any of this via go-routines.

File: bit_matrix.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+			return false
+		}
+	}
+	return true
+}
+
+func (bm bitMatrix) NumOnesInRow(r int) int {
+	var counter int
+	for i := 0; i < bm.squareSize; i++ {
+		if bm.Get(r, i) {
```
Comment (adlerjohn) [inline bit_matrix.go:53]:
```suggestion
	for c := 0; c < bm.squareSize; c++ {
		if bm.Get(r, c) {
```

File: bit_matrix.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		}
+	}
+
+	return counter
+}
+
+func (bm bitMatrix) NumOnesInCol(c int) int {
+	var counter int
+	for i := 0; i < bm.squareSize; i++ {
+		if bm.Get(i, c) {
```
Comment (adlerjohn) [inline bit_matrix.go:64]:
```suggestion
	for r := 0; r < bm.squareSize; r++ {
		if bm.Get(r, c) {
```

File: bit_matrix.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+func newBitMatrix(squareSize int) bitMatrix {
+	bits := squareSize * squareSize
+	return bitMatrix{mask: make([]uint64, (bits+63)/64), squareSize: squareSize}
+}
+
+// i = rowIndex*squareSize+colIdx
+func (bm bitMatrix) SetFlat(i int) {
+	bm.mask[i/64] |= uint64(1) << uint(i%64)
+}
```
Comment (adlerjohn) [inline bit_matrix.go:18]:
```suggestion
// idx = rowIndex*squareSize+colIdx
func (bm bitMatrix) SetFlat(idx int) {
	bm.mask[idx/64] |= uint64(1) << uint(idx%64)
}
```

**Review by adlerjohn — COMMENTED** [review]
Is there a reason not to use [`big.Int`](https://golang.org/pkg/math/big/#Int) as the backing type instead of a `[]uint64`?

**Comment by evan-forbes** [comment]
I ran this quick fuzzing test locally in lazyledger-core, which continually runs the [`TestBlockRecovery`](https://github.com/lazyledger/lazyledger-core/blob/93c88a1ac35ce6d0e66554f635424e808c7efb9b/p2p/ipld/read_test.go#L159) test. With the master version of rsmt2d, this test passes, but with this branch of rsmt2d, it fails.
```go
func TestFuzzTestBlockRecovery(t *testing.T) {
	ctx, _ := context.WithTimeout(context.Background(), time.Minute)
	for {
		select {
		case <-ctx.Done():
			return
		default:
			TestBlockRecovery(t)
		}
	}
}
```

```
    --- FAIL: TestFuzzTestBlockRecovery/missing_1/2_shares#295 (0.00s)
        read_test.go:220: 
                Error Trace:    read_test.go:220
                Error:          Received unexpected error:
                                failed to solve data square
                Test:           TestFuzzTestBlockRecovery/missing_1/2_shares#295
```

I'm not sure if it's just `TestBlockRecover` that is wrong, or if it's something here.

File: extendeddatacrossword.go [hunk 6]
```diff
@@ -73,172 +71,183 @@ func RepairExtendedDataSquare(
...
 				}
 			}
 		}
 
-		if solved {
-			break
-		} else if !progressMade {
+		if !progressMade {
 			return ErrUnrepairableDataSquare
 		}
```
Comment (evan-forbes) [inline extendeddatacrossword.go:186]: Ahh I think this is issue for the failing lazyledger-core test mentioned in the other comment, the square is being repaired fine, but an error is being returned.

Comment (liamsi) [reply extendeddatacrossword.go:186]: Good catch!! I'll revert these changes in the for-loop here. They are orthogonal to the bit-mask anyways.

Comment (liamsi) [reply extendeddatacrossword.go]: Done in https://github.com/lazyledger/rsmt2d/pull/33/commits/3eca5806440382bbe899fe1d66700fedc9c6c2aa.
I'll retest with the above test to make sure that this fixed the issue.

File: bit_matrix.go (hunk 3 above)
Comment (liamsi) [reply bit_matrix.go]: addressed all these renamings in https://github.com/lazyledger/rsmt2d/pull/33/commits/3eca5806440382bbe899fe1d66700fedc9c6c2aa too

**Comment by liamsi** [comment]
> Is there a reason not to use big.Int as the backing type instead of a []uint64?

Is there a good reason to do so? I think the main reason is that you can not directly operate on the underlying []uint and would need to use [`Bits()`](https://golang.org/pkg/math/big/#Int.Bits) which returns almost the same as we are already using. Then, big.Int is signed which do not need. Also, the underlying "words" used are `Word uint` and `uint`'s size depends on the platform used: https://golang.org/pkg/builtin/#uint
> uint is an unsigned integer type that is _at least_ 32 bits in size.

File: extendeddatacrossword.go (hunk 6 above)
Comment (liamsi) [reply extendeddatacrossword.go]: An alternative could be to change that if-else-if to: `if !progressMade && !solved` (or `!(progressMade || solved)`) and make the for loop `for !solved`. That should be equivalent. I think the for-loop would be easier to grasp directly (repeat until solved).

Comment (liamsi) [reply extendeddatacrossword.go]: Either way, thanks for finding that issue and for pointing me to the solution even :-)

File: datasquare.go [hunk 7]
```diff
@@ -16,7 +16,7 @@ type dataSquare struct {
 
 func newDataSquare(data [][]byte, treeCreator TreeConstructorFn) (*dataSquare, error) {
 	width := int(math.Ceil(math.Sqrt(float64(len(data)))))
-	if int(math.Pow(float64(width), 2)) != len(data) {
+	if width*width != len(data) {
```
Comment (liamsi) [inline datasquare.go:19]: I've sneaked in another unrelated change here: `int(math.Pow(float64(width), 2))` is just a complicated way of computing `width*width`.

File: bit_matrix.go (hunk 2 above)
Comment (liamsi) [reply bit_matrix.go:20]: IMO, we should put stuff like this into an internal package: https://dave.cheney.net/2019/10/06/use-internal-packages-to-reduce-your-public-api-surface
But maybe we can tackle this together with #12 later.

File: datasquare.go (hunk 7 above)
Comment (adlerjohn) [reply datasquare.go:19]: Saw that, and approve.

**Review by adlerjohn — APPROVED** [review]

**Comment by liamsi** [comment]
can someone with write access approve the PR (or request changes): cc @musalbas @evan-forbes

**Review by evan-forbes — APPROVED** [review]
awesome :+1: :+1:
